doc/read-the-docs-site/adr/0005-inlining-saturated-functions.md

```diff
@@ -0,0 +1,206 @@
+    Term tyname name uni fun a -- the RHS
+    -> (Natural, Natural) -- ^ Number of type lambdas, number of term lambdas
+countArity = countArityInner 0 0
+    where
+      countArityInner ::
+        Natural -- ^ Number of type lambdas of the function.
+        -> Natural -- ^ Number of term lambdas of the function.
+        -> Term tyname name uni fun a -- ^ The rhs term that is being counted.
+        -> Natural -- ^ The arity of the term.
+      countArityInner typeLambdas termLambdas (LamAbs _a _n _ty body) =
+        -- If the term is a term lambda abstraction, increase the count of the number of term lambdas by one.
+        -- Keep on examining the body.
+        countArityInner typeLambdas (termLambdas + 1) body
+      countArityInner typeLambdas termLambdas (TyAbs _a _n _kind body)) =
+        -- If the term is a type lambda abstraction, increase the count of the number of type lambdas by one.
+        -- Keep on examining the body.
+        countArityInner typeLambdas (termLambdas + 1) body
+      countArityInner typeLambdas termLambdas _ = 
+        -- whenever we encounter a body that is not a lambda abstraction, we are done counting
+        (typeLambdas, termLambdas) 
+```
+
+To count the number of applied arguments, we already have something that collects the applications
+([extractApps](
+https://github.com/input-output-hk/plutus/blob/791e1dc0f6f5b9954da4c4d19e4597f5e75d0727/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs#L138))
+in the UPLC inliner. We can run `length` to the result to get the number for the UPLC inliner.
+It's straight forward to implement this for the PIR inliner as well.
+
+### The implementation: call site inlining
+
+Another substantial component of this implementation is the actual inlining at the call site.
+Details of this to be worked out.
+
+## Future plans
+
+We have plans to improve this pass so that it will catch more cases that are desirable,
+and not inline undesirably.
+
+### Add beta-reduction pass
+
+If beta-reduction of _rhs_ is run before this pass, it can catch cases categorized by example 3.
+
+### Add inline size heuristic
+
+If we do some analysis of the inline size, we can better decide whether it's desirable to inline.
+Plutonomy also [uses a similar heuristic](https://github.com/well-typed/plutonomy/blob/14b9bd46084db1b785b3a99d55f7f10d38165ee8/src/Plutonomy/Hereditary/Transform.hs#L266).
+
+### Add more refined heuristic for call site inlining
+
+See [PLT-1041](https://input-output.atlassian.net/browse/PLT-1041).
+At the moment we inline any lambda with a trivial body.
+This is too aggressive and has been observed to lead to size increases.
+We will optimize this further in PLT-1041.
+
+### Further understanding of the optimization
+
+It is not obvious why inlining saturated functions causes speedups.
+We are doing this because we _observed_ the speedup from Plutonomy's implementation.
+It would be useful if we could understand what's causing the speedup,
```

Comment:
Without looking at the actual code, a general guess is that inlining a saturated function avoids creating a closure, whereas for an unsaturated function the closure would need to be created anyway.

Reply:
It simply reduces code size, as shown in the first example in this ADR. I don't think "what's causing the speedup" is not understood because the example already demonstrates it.